pkg/http/server.go

```diff
@@ -0,0 +1,233 @@
+package http
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options/util"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+	"golang.org/x/sync/errgroup"
+)
+
+// Server represents an HTTP or HTTPS server.
+type Server interface {
+	// Start blocks and runs the server.
+	Start(ctx context.Context) error
```

Comment:
Do we want to expose `Shutdown` (or `Close`) on this interface?

Reply:
The idea I'm running with on this one is that the context being cancelled causes the shutdown, so to use it you do 

```
ctx, cancel := context.WithCancel(context.Background())

go srv.Start(ctx)

// wait for signal
cancel()
```

If you think having a Close would help then I can look into adding that as well, would probably mean just moving the context inside the server itself implementation wise.

Reply:
I gotcha -- so this `ctx` in the Server `Start` is more for controlling the concurrency with the underlying uses of the `errgroup`. (rather than in something like `http.Server` which doesn't take it).

I think I had `Serve` + `Shutdown`/`Close` on my mind when I saw an interface named `Server` (and mistook `Start` for `Serve`).

Reply:
Yep that's it, this is more of a Start with a stop channel style thing, except the stop channel is the context which can be cancelled